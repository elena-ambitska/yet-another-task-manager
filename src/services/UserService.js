import BaseService from "./BaseService";

class UserService extends BaseService {

    async register(data) {
        return this.post(`auth/local/register`, data);
    }

    async login(data) {
        return this.post(`auth/local`, data);
    }
}

export default new UserService();