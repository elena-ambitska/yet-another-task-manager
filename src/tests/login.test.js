import React from "react";
import Login from "../components/login";
import toJson from "enzyme-to-json";
import UserService from "../services/UserService";
import resultLoginMock from "./mock/login";
import resultLoginFailMock from "./mock/login_fail_400";
import {act, fireEvent, screen} from "@testing-library/react";

import {renderWithContext} from "./render";

const getUserServiceDataMock = jest.spyOn(UserService, "login");

beforeEach(() => {
    getUserServiceDataMock.mockClear();
});

const mockedUsedNavigate = jest.fn();


jest.mock('react-router', () => ({
    ...jest.requireActual('react-router'),
    useNavigate: () => mockedUsedNavigate,
}));
const loginMock = (result) => {
    return async (dispatch) => {
        return Promise.resolve(result);
    };
};

describe("Test Login component", () => {

    it("should render login component",  () => {
        renderWithContext(<Login/>);

        expect(screen.getByLabelText("Email")).toBeVisible()
        expect(screen.getByLabelText("Password")).toBeVisible();
    });

    it("login component sends data to services", async  () => {
        getUserServiceDataMock.mockReturnValueOnce(loginMock(resultLoginMock));
        renderWithContext(<Login/>)

        await act(() => {
            fireEvent.submit(screen.getByRole("form-submit"));
        })

        expect(getUserServiceDataMock).toHaveBeenCalledTimes(1);
    });

    it("login component displays error when fails", async () => {
        getUserServiceDataMock.mockReturnValueOnce(loginMock(resultLoginFailMock));
        renderWithContext(<Login/>)

        await act(() => {
            fireEvent.submit(screen.getByRole("form-submit"));
        })
        expect(getUserServiceDataMock).toHaveBeenCalledTimes(1);
        const errorList =  await screen.getByRole("error-list");
        expect(errorList.children.length).toBe(resultLoginFailMock.data[0].messages.length);
        expect(errorList.children[0]).toHaveTextContent(resultLoginFailMock.data[0].messages[0].message);
    });
});
