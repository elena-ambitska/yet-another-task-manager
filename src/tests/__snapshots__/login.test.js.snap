// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Login component should render login component 1`] = `
<input
  class="form-control"
  id="loginInput"
  name="identifier"
  required=""
  type="text"
  value=""
/>
`;

exports[`Test Login component should render login component 2`] = `
<input
  class="form-control"
  id="passwordInput"
  name="password"
  required=""
  type="password"
  value=""
/>
`;
